core/shared/src/main/scala/fs2/retry.scala

```diff
@@ -0,0 +1,57 @@
+package fs2
+
+import cats.effect.Async
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+object retry {
+  /**
+   * Retries `fa` on failure, returning a single element
+   * stream with the result of `fa` as soon as it succeeds.
+   *
+   * @param delay Duration of delay before the first retry
+   *
+   * @param nextDelay Applied to the previous delay to compute the
+   *                  next, e.g. to implement exponential backoff
+   *
+   * @param maxRetries Number of attempts before failing with the
+   *                   latest error, if `fa` never succeeds
+   *
+   * @param retriable Function to determine whether a failure is
+   *                  retriable or not, defaults to retry every
+   *                  `NonFatal`. A failed stream is immediately
+   *                  returned when a non-retriable failure is
+   *                  encountered
+   */
+  def retry[F[_], A](fa: F[A],
+                     delay: FiniteDuration,
+                     nextDelay: FiniteDuration => FiniteDuration,
+                     maxRetries: Int,
+                     retriable: Throwable => Boolean = NonFatal.apply)(
+    implicit F: Async[F], scheduler: Scheduler, ec: ExecutionContext): Stream[F, A] = {
+     val delays = Stream.unfold(delay)(d => Some(d -> nextDelay(d))).covary[F]
+
+    retries(fa, delays)
+      .take(maxRetries)
+      .takeThrough(_.fold(err => retriable(err), _ => false))
+      .last
+      .map(_.get) // `retries` is guaranteed to have at least one element
```

Comment:
I'd rather do something like `.getOrElse(sys.error("[fs2] impossible: empty stream in `retry`"))`